For code review

JKuskys opened the pull request.

mildapetr committed:
feat: working user rest controller with gets only

Rytautas committed:
added application component
added sql table
fixed bugs between user component and sql table

mildapetr committed:
feat: full user rest controller (without validation)

mildapetr committed:
Merge remote-tracking branch 'origin/master'

Rytautas committed:
added get, post, put, delete methods for application component
minor bug fixed for user component

Rytautas committed:
patch: replaced name and last_name fields with full_name field in application component and database

mildapetr committed:
feature/user_validation (#1)

* adding mandatory field validation for user fields

* adding user email field validation

* adding user password validation

* updating error messages to lithuanian

Rytautas committed:
application_validation (#2)

* changed size for phone number in database from 255 to 20 characters

* added validation application

* added validation exception handler for application component

mildapetr committed:
feature/user_info_handling (#3)

* adding existing email exceptions with better exception handling

* updating ids to be auto incremented

mildapetr committed:
fixed application submission missing user validation (#4)

mildapetr committed:
fix/registration_errors (#5)

* fixed id and error issues

* simplifying update by not allowing to update values (except for admin review status later)

* fix/minor changes

* minor changes #2

Co-authored-by: Rytautas <rytautas.kvasinskas@gmail.com>

mildapetr committed:
test/user_validation (#6)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

Rytautas committed:
Security (#9)

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* fixed issue with POST request for /auth/login being denied. it now returns an internal server error
minor changes to DB that will need to be fixed later

* changed autowiring to go on constructor and not on field

* some changes

* change session duration to 15min

* fix/registration_errors (#5)

* fixed id and error issues

* simplifying update by not allowing to update values (except for admin review status later)

* fix/minor changes

* minor changes #2

Co-authored-by: Rytautas <rytautas.kvasinskas@gmail.com>

* test/user_validation (#6)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* fix: fix user authentication and password encryption

* fixed REST controllers returning too much information when creating and updating users/applications

Co-authored-by: mildapetr <58117403+mildapetr@users.noreply.github.com>
Co-authored-by: gaijinn <gaijinn@users.noreply.github.com>

Rytautas committed:
database changes for the upcoming features (#8)

Rytautas committed:
Security fixes (#10)

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* fixed issue with POST request for /auth/login being denied. it now returns an internal server error
minor changes to DB that will need to be fixed later

* changed autowiring to go on constructor and not on field

* some changes

* change session duration to 15min

* fix/registration_errors (#5)

* fixed id and error issues

* simplifying update by not allowing to update values (except for admin review status later)

* fix/minor changes

* minor changes #2

Co-authored-by: Rytautas <rytautas.kvasinskas@gmail.com>

* test/user_validation (#6)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* fix: fix user authentication and password encryption

* fixed REST controllers returning too much information when creating and updating users/applications

* Security (#9)

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* changed username to email as we are using email instead of username

* fixed bean autowiring issue

* fixed issue with POST request for /auth/login being denied. it now returns an internal server error
minor changes to DB that will need to be fixed later

* changed autowiring to go on constructor and not on field

* some changes

* change session duration to 15min

* fix/registration_errors (#5)

* fixed id and error issues

* simplifying update by not allowing to update values (except for admin review status later)

* fix/minor changes

* minor changes #2

Co-authored-by: Rytautas <rytautas.kvasinskas@gmail.com>

* test/user_validation (#6)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* added authentication base with jwt token
changes to database (user is_admin field replaced with roles field)
model/user now implements UserDetails interface
added model/AuthenticationRequest model

bugs:
for some reason denies access to /auth/login POST method even though its permitted
does not load up applications

* fix: fix user authentication and password encryption

* fixed REST controllers returning too much information when creating and updating users/applications

Co-authored-by: mildapetr <58117403+mildapetr@users.noreply.github.com>
Co-authored-by: gaijinn <gaijinn@users.noreply.github.com>

* database changes for the upcoming features (#8)

* formatted code

* removed unnecessary imports

* fixed warnings with ResponseEntity usage

* fixed issues that appeared after merging security branch
made some changes to the database

* added encrypted passwords to database

* removed the unnecessary assignment in UserController class
renamed getUsers to fetchAllUsers in UserController class
renamed newUser to user in UserController class
made corresponding changes to ApplicationController class
added curly braces {} for if statements in UserServiceImpl and ApplicationServiceImpl classes

renamed DemoApplication class to ITAcademyBackend
changed the names of all variables in Application class to camel case
changed several fields' names in Application class to better reflect what values are held in them
removed unnecessary comments

Co-authored-by: mildapetr <58117403+mildapetr@users.noreply.github.com>
Co-authored-by: gaijinn <gaijinn@users.noreply.github.com>

Rytautas committed:
allowing access to api/applications post method for registration

JKuskys committed:
Merge pull request #13 from JKuskys/fix/security-permissions

allowing access to api/applications post method for registration

Rytautas committed:
allowing access to api/applications post method for registration

JKuskys committed:
Merge pull request #14 from JKuskys/naming-and-typing-changes

naming and typing changes

Rytautas committed:
Feat/application states and comments (#15)

* allowing access to api/applications post method for registration

* added admin comment component with basic REST implementations that need to be remade

* added admin comment component with basic REST implementations that need to be remade

Rytautas committed:
Feat/application states and comments (#16)

* added admin comment component with basic REST implementations that need to be remade

* added admin comment component with basic REST implementations that need to be remade

* changed the format of data returned by applicationController GET method to return less unnecessary data
minor changes to database

ModestaC committed:
Added functionality for adding and modifying new comment (#17)

Co-authored-by: Modesta <m.cirtautaite@gmail.com>

ModestaC committed:
Feat/comment update and create (#18)

* Added functionality for adding and modifying new comment

* Added functionality for adding and modifying new comment

Co-authored-by: Modesta <m.cirtautaite@gmail.com>

grkarolis commented on src/main/java/com/project/controller/AdminCommentController.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<HttpStatus> createComment(@Valid @RequestBody RequestComment comment) throws ApplicationNotFoundException, UserNotFoundException {
```
Word `Request` should be the last one in a class name. Something like `FetchCustomersRequest`.

Rytautas replied:
renamed RequestComment class to CommentRequest

grkarolis commented on src/main/java/com/project/controller/AdminCommentController.java:
```diff
@@ -0,0 +1,80 @@
+    @PutMapping("/{id}")
+    public ResponseEntity<HttpStatus> updateComment(@RequestBody RequestComment comment, @PathVariable Long id)
+            throws AdminCommentNotFoundException, ApplicationNotFoundException, UserNotFoundException {
+        AdminComment adminComment = new AdminComment(comment.getComment(), comment.getDate(),
```
`AdminComment` sounds very specific. I would suggest creating an Interface or an Abstract class Comment and then extend or implement them for `AdminComment` since if in the future you will need to have let's say `UserComment` you will just need to extend or implement `Comment` and add User-specific data.

Rytautas replied:
renamed `AdminComment` and all related classes to `Comment`, to be less specific

grkarolis commented on src/main/java/com/project/controller/AdminCommentController.java:
```diff
@@ -0,0 +1,80 @@
+@RestController
+@CrossOrigin(origins = "*")
+@RequestMapping("api/comments")
+public class AdminCommentController {
```
If I am not mistaken your API name is `"api/comments"` to fetch comment related resources. I would maybe suggest structuring your path to the comment resource like this: `"/api/applications/{applicationId}/comments"` because your comments are part of your application. When consuming your API it will be easier to understand where the comments come from.

Rytautas replied:
changed path to `"/api/applications/{appId}/comments"`, the controller now works for specific application and not all comments

grkarolis commented on src/main/java/com/project/controller/ApplicationController.java:
```diff
@@ -0,0 +1,72 @@
+        List<Application> applications = applicationService.getAll();
+        List<ApplicationResponse> response = new ArrayList<>();
+
+        for (Application app : applications) {
```
Here you could use Java 8 map to simpligy and optimize your code.
https://mkyong.com/java8/java-8-streams-map-examples/

Rytautas replied:
replaced for each loops with java 8 stream map

grkarolis commented on src/main/java/com/project/controller/ApplicationController.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @PostMapping
+    ResponseEntity<HttpStatus> createApplication(@Valid @RequestBody Application application) throws UserException, ApplicationNotFoundException {
```
I would suggest creating `ApplicationRequest` class and not to use the same classes you use to create your DB structure.

Rytautas replied:
created response entities for ApplicationRequest, AuthenticationRequest, CommentRequest, ApplicationResponse, AuthenticationResponse, CommentResponse and UserResponse to reduce the ammount of information moving through controllers between front end and back end

grkarolis commented on src/main/java/com/project/controller/AuthenticationController.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @PostMapping("/login")
+    public ResponseEntity<Map<Object, Object>> login(@RequestBody AuthenticationRequest data) {
```
I suggest creating something like `AuthenticationResponse` and returning it. Also, I don't think you need to return the user's email. The token should be sufficient enough unless there is some kind of requirement for it.

Rytautas replied:
created response entities for ApplicationRequest, AuthenticationRequest, CommentRequest, ApplicationResponse, AuthenticationResponse, CommentResponse and UserResponse to reduce the ammount of information moving through controllers between front end and back end

grkarolis commented on src/main/java/com/project/exception/AdminCommentNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class AdminCommentNotFoundException extends Exception {
+    public AdminCommentNotFoundException(long id) {
+        super("Komentaras su nurodytu identifikatoriumi (" + id + ") neegzistuoja.");
```
You could use `super(String.format("Komentaras su nurodytu identifikatoriumi (%d) neegzistuoja.", id))`. Looks a bit cleaner.

Rytautas replied:
changed validation messages to use `String.format()`

grkarolis commented on src/main/java/com/project/model/request/RequestComment.java:
```diff
@@ -0,0 +1,31 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class RequestComment {
```
No data validation?

Rytautas replied:
moved field validation from database entity models to request entity models

grkarolis commented on src/main/java/com/project/model/response/ApplicationResponse.java:
```diff
@@ -0,0 +1,43 @@
+
+    private int commentCount;
+
+    private String status;
```
Can status be an `Enumeration` if there is a predefined set of statuses?

Rytautas replied:
changed status type to an Enumeration of predefined set of statuses

Rytautas committed:
limited the information sent and required for comment REST methods (#19)

Rytautas committed:
naming change

Rytautas committed:
naming change

Rytautas committed:
naming and typing changes
changed application status to enum
moved fullName field from application to user
limited information provided by REST controllers
limited information required by REST controllers

Rytautas committed:
Merge branch 'naming-and-typing-changes'

Rytautas committed:
Naming and typing changes (#20)

* limited the information sent and required for comment REST methods

* naming change

* naming change

* naming and typing changes
changed application status to enum
moved fullName field from application to user
limited information provided by REST controllers
limited information required by REST controllers

Rytautas committed:
added authenticationResponse model (#21)

changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

Rytautas committed:
Validation and exception changes (#22)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* changed naming of a few fields in model/Application class (from application_date to applicationDate etc.)

* added authenticationResponse model
changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

* moved value validation from database entity models to request entity models
changed exception and validation messages to use String.format()

Co-authored-by: Milda Petrikaite <Petrikaite.M@gmail.com>

Rytautas committed:
Validation and exception changes (#23)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* changed naming of a few fields in model/Application class (from application_date to applicationDate etc.)

* added authenticationResponse model
changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

* moved value validation from database entity models to request entity models
changed exception and validation messages to use String.format()

* fixed requiring comment author fullName and email to post a new comment

Co-authored-by: Milda Petrikaite <Petrikaite.M@gmail.com>

Rytautas committed:
Validation and exception changes (#24)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* changed naming of a few fields in model/Application class (from application_date to applicationDate etc.)

* added authenticationResponse model
changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

* moved value validation from database entity models to request entity models
changed exception and validation messages to use String.format()

* fixed requiring comment author fullName and email to post a new comment

* fixed requiring comment author fullName and email to post a new comment

Co-authored-by: Milda Petrikaite <Petrikaite.M@gmail.com>

grkarolis commented on src/main/java/com/project/controller/AdminCommentController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<CommentResponse>> fetchComments(@PathVariable("appId") long appId) {
```
Just a "small" `long`?

Rytautas replied:
changed type `long` values to type `Long`

grkarolis commented on src/main/java/com/project/controller/AdminCommentController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @PostMapping
+    public ResponseEntity<CommentResponse> createComment(@Valid @RequestBody CommentRequest comment, @PathVariable("appId") long appId)
```
A "small" `long` here as well?

Rytautas replied:
changed type `long` values to type `Long`

grkarolis commented on src/main/java/com/project/controller/AdminCommentController.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @PutMapping("/{id}")
+    public ResponseEntity<CommentResponse> updateComment(@RequestBody CommentRequest comment, @PathVariable("id") Long id, @PathVariable("appId") long appId)
```
A "small" `long` here as well?

grkarolis commented on src/main/java/com/project/controller/ApplicationController.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<ApplicationResponse> fetchApplication(@PathVariable long id) throws ApplicationNotFoundException {
```
A "small" `long` here as well? You should be consistent with your variables.

Rytautas replied:
changed type `long` values to type `Long`

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @GetMapping("/{id}")
+    ResponseEntity<UserResponse> fetchUser(@PathVariable long id) throws UserNotFoundException {
```
"Small" long?

Rytautas replied:
changed type `long` values to type `Long`

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @DeleteMapping("/{id}")
+    ResponseEntity<HttpStatus> deleteUser(@PathVariable Long id) throws UserNotFoundException {
```
"Big" long?

Rytautas replied:
changed type `long` values to type `Long`

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @PostMapping
+    ResponseEntity<HttpStatus> createUser(@RequestBody User user) throws UserException {
```
If this class is only used for requests maybe it would be convenient to name it something like `UserRequest` if your order classes used for requests ends with 'Request'?

Rytautas replied:
Added model entity for `UserRequest`

grkarolis commented on src/main/java/com/project/model/request/ApplicationRequest.java:
```diff
@@ -0,0 +1,54 @@
+    @NotBlank(message = "Paraiškos data negali būti tuščia")
+    private String applicationDate;
+
+    private User user;
```
Just a thought but maybe an `Applicant` would sound better rather then `User` in the application?

Rytautas replied:
renamed the field to `User applicant`. Is this what you meant by the comment or did you mean to change to `Applicant applicant`?

grkarolis commented on src/main/java/com/project/security/JwtTokenProvider.java:
```diff
@@ -0,0 +1,90 @@
+@Component
+public class JwtTokenProvider {
+    @Value("${security.jwt.token.secret-key:secret}")
+    private String secretKey = "itacademy2020";
```
You could move out this value to a property file so you won't need to expose it directly in the code.

Rytautas replied:
removed the "hard-coded" values

grkarolis commented on src/main/java/com/project/security/JwtTokenProvider.java:
```diff
@@ -0,0 +1,90 @@
+    private String secretKey = "itacademy2020";
+
+    @Value("${security.jwt.token.expire-length:900000}")
+    private long validityInMilliseconds = 900000; // 15min sesijos trukme
```
You could move out this value to a property file so you won't need to expose it directly in the code.

Rytautas replied:
removed the "hard-coded" values

grkarolis commented on src/main/java/com/project/service/AdminCommentServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    public AdminComment addAdminComment(AdminComment adminComment) {
+        adminComment.setId(null);
```
Why do you need this?

Rytautas replied:
reworked the `ServiceImpl` classes so that `null` values are no longer assigned

grkarolis commented on src/main/java/com/project/service/ApplicationServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+                    existingApp.setStatus(application.getStatus());
+                    return applicationRepository.save(existingApp);
+                })
+                .orElse(null);
```
Maybe change to `Optional`?

Rytautas replied:
reworked the `ServiceImpl` classes so that `null` values are no longer returned

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+            throw new UserEmailExistsException(user.getEmail());
+        }
+
+        user.setId(null);
```
Why  you need to set `null`?

Rytautas replied:
reworked the `ServiceImpl` classes so that `null` values are no longer assigned

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+        user.setPassword(bcryptEncoder.encode(user.getPassword()));
+
+        List<String> roles = new ArrayList<>();
+        roles.add("USER");
```
Maybe you could have `Enumeration` for your roles? Easier to keep track of them when they are in one place and not harcoded everywhere.

Rytautas replied:
created `Enumeration` for roles

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+                    existingUser.setFullName(user.getFullName());
+                    return userRepository.save(existingUser);
+                })
+                .orElse(null);
```
Maybe `Optional`?

Rytautas replied:
reworked the `ServiceImpl` classes so that `null` values are no longer returned

grkarolis commented on src/main/java/com/project/validation/MandatoryValueValidator.java:
```diff
@@ -0,0 +1,11 @@
+public class MandatoryValueValidator extends Validator<String> {
+    @Override
+    public void validate(String attribute, String message) {
+        if(attribute == null || attribute.trim().isEmpty())
```
You could use apache commons lib here `StringUtils.isEmpty()` checks whether `String` contains text and is `null` safe.

You can read more about it here:
https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html

Rytautas replied:
changed the `if` statement condition to `StringUtils.isEmpty()` method

Rytautas committed:
Validation and exception changes (#25)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* changed naming of a few fields in model/Application class (from application_date to applicationDate etc.)

* added authenticationResponse model
changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

* moved value validation from database entity models to request entity models
changed exception and validation messages to use String.format()

* fixed requiring comment author fullName and email to post a new comment

* fixed requiring comment author fullName and email to post a new comment

* getting a users application by email

* fixed validation bug

Co-authored-by: Milda Petrikaite <Petrikaite.M@gmail.com>

mildapetr committed:
adding some more believable testing data (#26)

Rytautas committed:
Feat/user application get (#27)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* changed naming of a few fields in model/Application class (from application_date to applicationDate etc.)

* added authenticationResponse model
changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

* moved value validation from database entity models to request entity models
changed exception and validation messages to use String.format()

* fixed requiring comment author fullName and email to post a new comment

* fixed requiring comment author fullName and email to post a new comment

* getting a users application by email

* fixed validation bug

* Added functionality to fetch specific users application by email

Co-authored-by: Milda Petrikaite <Petrikaite.M@gmail.com>

Rytautas committed:
Feat/user application get (#28)

* adding email validation tests

* adding password validation tests

* adding mandatory field validation tests

* adding full user validation tests

* changed naming of a few fields in model/Application class (from application_date to applicationDate etc.)

* added authenticationResponse model
changed comment REST controller paths to include application id
comment REST controllers now work for specific applications
replaced for each loops with more efficient stream mapping

* moved value validation from database entity models to request entity models
changed exception and validation messages to use String.format()

* fixed requiring comment author fullName and email to post a new comment

* fixed requiring comment author fullName and email to post a new comment

* getting a users application by email

* fixed validation bug

* Added functionality to fetch specific users application by email

* changed from GET method to POST

Co-authored-by: Milda Petrikaite <Petrikaite.M@gmail.com>

Rytautas committed:
Naming and typing changes (#29)

* limited the information sent and required for comment REST methods

* naming change

* naming change

* naming and typing changes
changed application status to enum
moved fullName field from application to user
limited information provided by REST controllers
limited information required by REST controllers

* changed id types from long to Long for consistency

* created new model UserRequest to control the amount of information needed to create a new user
created new model ApplicationUpdateRequest to reduce the amount of information needed to update the status of request

* renamed AdminComment to Comment (controllers, repositories and services included)
moved the logic from CommentController to CommentServiceImpl
moved the logic from ApplicationController to ApplicationServiceImpl
moved the logic from UserController to UserService

* renamed AdminCommentNotFound exception to CommentNotFound

* added enum for user roles
removed hard-coded values
removed some unnecessary imports

grkarolis commented on src/main/java/com/project/controller/AuthenticationController.java:
```diff
@@ -0,0 +1,53 @@
+            String token = jwtTokenProvider.createToken(email, this.users.findByEmail(email)
+                    .orElseThrow(() -> new UsernameNotFoundException("Neteisingas vartotojo paštas arba slaptažodis"))
+                    .getRoles());
+            AuthenticationResponse model = new AuthenticationResponse(email, token);
```
Why this is called `model`? Maybe it would be better to name it `response`?

Rytautas replied:
renamed variable from `model` to `response`

grkarolis commented on src/main/java/com/project/controller/CommentController.java:
```diff
@@ -0,0 +1,57 @@
+    @PostMapping
+    public ResponseEntity<CommentResponse> createComment(@Valid @RequestBody CommentRequest comment, @PathVariable("appId") Long appId)
+            throws ApplicationNotFoundException, UserNotFoundException {
+
```
Is this empty line necessary?

Rytautas replied:
removed unnecessary empty lines

grkarolis commented on src/main/java/com/project/controller/CommentController.java:
```diff
@@ -0,0 +1,57 @@
+    @PutMapping("/{id}")
+    public ResponseEntity<CommentResponse> updateComment(@RequestBody CommentRequest comment, @PathVariable("id") Long id, @PathVariable("appId") Long appId)
+            throws CommentNotFoundException {
+
```
Is this empty line necessary?

Rytautas replied:
removed unnecessary empty lines

grkarolis commented on src/main/java/com/project/controller/UserController.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @PostMapping("/application")
+    public ResponseEntity<ApplicationResponse> fetchUserApplication(@RequestBody UserCommentRequest email) throws UserNotFoundException {
```
Why this service fetches application in a user controller also why it has a `UserCommentRequest` as request but not the Comment is fetched from the service instead it fetches the Application?

Rytautas replied:
renamed `UserCommentRequest` to `UserApplicationRequest` to better reflect what it is used for. As for why does this method fetch an application in the user controller, it is because it uses the same request mapping path. Do you think it is better to move this to a new controller?

grkarolis commented on src/main/java/com/project/model/User.java:
```diff
@@ -0,0 +1,90 @@
+        this.fullName = user.getFullName();
+        this.email = user.getEmail();
+        this.password = user.getPassword();
+        this.roles =  new ArrayList<>(Collections.singletonList(UserRoles.USER.toString()));
```
You can use the method `UserRoles.USER.name()` to use the text of the enum constant.

Rytautas replied:
replaced `UserRole.USER.toString()` methods to `UserRole.USER.name()

grkarolis commented on src/main/java/com/project/model/UserRoles.java:
```diff
@@ -0,0 +1,6 @@
+package com.project.model;
+
+public enum UserRoles {
```
Your enum should be named `UserRole` since when you use enum you take a singular value for instance `ADMIN`.

Rytautas replied:
renamed enumeration to `UserRole`

Rytautas committed:
fixed bug with new applications not getting saved in the database (#30)

Rytautas committed:
Naming and typing changes (#31)

* limited the information sent and required for comment REST methods

* naming change

* naming change

* naming and typing changes
changed application status to enum
moved fullName field from application to user
limited information provided by REST controllers
limited information required by REST controllers

* changed id types from long to Long for consistency

* created new model UserRequest to control the amount of information needed to create a new user
created new model ApplicationUpdateRequest to reduce the amount of information needed to update the status of request

* renamed AdminComment to Comment (controllers, repositories and services included)
moved the logic from CommentController to CommentServiceImpl
moved the logic from ApplicationController to ApplicationServiceImpl
moved the logic from UserController to UserService

* renamed AdminCommentNotFound exception to CommentNotFound

* added enum for user roles
removed hard-coded values
removed some unnecessary imports

* several naming changes

* several naming changes

Rytautas committed:
changes to database, models and controllers to save comment status isVisibleToApplicant for comments visible to applicants (#32)

Rytautas committed:
Feat/student comments (#33)

* changes to database, models and controllers to save comment status isVisibleToApplicant for comments visible to applicants

* changed database to hold datetime on applications and comments
creating new application/comment no longer requires dates

justas68 commented on src/main/java/com/project/model/User.java:
```diff
@@ -0,0 +1,90 @@
+        return true;
+    }
+
+    public User (UserRequest user) {
```
constructor should be before methods

Rytautas replied:
moved constructor before methods

justas68 commented on src/main/java/com/project/model/User.java:
```diff
@@ -0,0 +1,90 @@
+        this.fullName = user.getFullName();
+        this.email = user.getEmail();
+        this.password = user.getPassword();
+        this.roles =  new ArrayList<>(Collections.singletonList(UserRole.USER.name()));
```
Can use List.of(UserRole.USER.name()) here

justas68 replied:
or just Collections.singletonList(UserRole.USER.name()) if you like it better, but no need to create new ArrayList

Rytautas replied:
changed to `Collections.singletonList(UserRole.USER.name())`

justas68 commented on src/main/java/com/project/service/ApplicationServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public void deleteApplication(Long id) throws ApplicationNotFoundException {
+        if (!applicationRepository.findById(id).isPresent()) {
```
could create jpa method to check if application exists, would not need to retrieve it then

Rytautas replied:
replaced several calls to repository with `existsById()`

Rytautas committed:
Fix/bug fixes (#34)

* fixed bug with new applications not getting saved in the database

* Merge branch 'fix/bug-fixes' of D:\Programming\ITAcademy with conflicts.

* code cleanup

Rytautas committed:
Validation/validation changes (#35)

* fixed bug with new applications not getting saved in the database

* Merge branch 'fix/bug-fixes' of D:\Programming\ITAcademy with conflicts.

* code cleanup

* created custom annotation for validating password
added @Valid annotation on fields where it was necessary and missed

Rytautas committed:
Fix/bug fixes (#36)

* fixed bug with new applications not getting saved in the database

* Merge branch 'fix/bug-fixes' of D:\Programming\ITAcademy with conflicts.

* code cleanup

* fixed application PUT method failing validation and returning 500 code

grkarolis commented on src/main/java/com/project/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public CommentResponse addAdminComment(CommentRequest comment, Long appId)
+            throws ApplicationNotFoundException, UserNotFoundException {
+        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
You use the same `DateFormat` multiple times. Maybe you could have just only one instance of it?

grkarolis commented on src/main/java/com/project/model/Application.java:
```diff
@@ -0,0 +1,82 @@
+    private User applicant;
+
+    public Application(ApplicationRequest app, ApplicationStatus status, List<Comment> comments, User applicant) {
+        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
It is not a good practice to format data in a constructor.

Rytautas replied:
data formatting removed from constructor

mildapetr committed:
feature/IT11.2 (#37)

* adding comment attachment fetching endpoint

* fixing some previous issues and adding file upload logic

mildapetr committed:
adding a possible fix to file path problem (#38)

Rytautas committed:
UNIT TESTS!!! (#39)

minor changes to controller classes to allow testing
unit tests for application/comment/user controller classes
unit tests for application/comment/user service classes

mildapetr committed:
adding another possible solution (#40)

mildapetr committed:
adding final solution to the attachment problem (I hope this finally works god dammit) (#41)

Rytautas committed:
Feat/notifications db changes (#42)

* changes to db to store notification details
changes to application models and comment service to store and retrieve notification details

* changed default values in database for bool from null to false

* same as last, except actually doing it

Rytautas committed:
Feat/notifications db changes (#43)

* changes to db to store notification details
changes to application models and comment service to store and retrieve notification details

* changed default values in database for bool from null to false

* same as last, except actually doing it

* resetting new notification booleans when getting application

Rytautas committed:
Feat/notifications db changes (#44)

* changes to db to store notification details
changes to application models and comment service to store and retrieve notification details

* changed default values in database for bool from null to false

* same as last, except actually doing it

* resetting new notification booleans when getting application

* maybe fixed resetting booleans for notifications

Rytautas committed:
Feat/notifications db changes (#45)

* changes to db to store notification details
changes to application models and comment service to store and retrieve notification details

* changed default values in database for bool from null to false

* same as last, except actually doing it

* resetting new notification booleans when getting application

* maybe fixed resetting booleans for notifications

* added email service
new comments to student from admin now send email to student
status updates to PRIIMTA and ATMESTA now send email to student

mildapetr committed:
adding file upload validation (#46)

mildapetr committed:
making small adjustment to accepted content types (#47)

Rytautas committed:
Feat/password reset (#48)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

grkarolis commented on src/main/java/com/project/controller/CommentController.java:
```diff
@@ -0,0 +1,77 @@
+            throws CommentNotFoundException, CommentAttachmentNotFoundException {
+        byte[] data = commentService.getAttachment(id, filename);
+        return ResponseEntity.ok().header(HttpHeaders.CONTENT_DISPOSITION,
+                "attachment; filename=\"" + filename + "\"").body(data);
```
Maybe you could use `String.format("attachment; filename=\%s\", filename)`;

Rytautas replied:
replaced string addition to `String.format()`

grkarolis commented on src/main/java/com/project/controller/CommentController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @PostMapping("/{id}/attachment")
+    public ResponseEntity<HttpStatus> createCommentAttachment(@PathVariable("id") Long id, @RequestParam("file")MultipartFile file)
```
Why file is a request param and not a request body?

mildapetr replied:
because file has some metadata and other things so that it could be processed as multipart file, which comes from multipart/form-data. Researched it a lot, and people on the internet mainly suggest just posting the file through form data because otherwise some additional binary actions are needed and it also increases the size of the file and the complexity of processing it. If I use multipart file, all these are handled by default implementations, which work well and I don't really feel brave enough to challenge them

grkarolis commented on src/main/java/com/project/controller/CommentController.java:
```diff
@@ -0,0 +1,77 @@
+        return new ResponseEntity<>(commentService.getById(id), HttpStatus.OK);
+    }
+
+    @GetMapping(value = "/{id}/attachment/{filename}")
```
Why you chose `filename` as a path variable and not a file ID?

mildapetr replied:
well, as we can have only one file attachment per comment, it was convenient to make a field for it in the database rather than make a new table for files, which leaves files with no id. Another problem that I had was that if I didn't end the path with filename, it would not understand that it is the file that I uploaded and would download some nonsense file with no extension. I might have made a mistake somewhere while doing that, but when I changed to filename, it worked well, so I left it this way.

grkarolis commented on src/main/java/com/project/controller/PasswordResetController.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @PostMapping("/resetPassword")
+    public GenericResponse resetPassword(HttpServletRequest request, @RequestParam("email") String userEmail) throws UserNotFoundException {
```
You need to have a request body for POST methods.

Rytautas replied:
it's OK to send a POST request without a body and instead use query string parameters.

grkarolis commented on src/main/java/com/project/controller/PasswordResetController.java:
```diff
@@ -0,0 +1,65 @@
+        return "redirect:/updatePassword";
+    }
+
+    @PostMapping("/resetPassword")
```
Please rename your URI to something like `reset-password` because this does not comply with Open API guidelines.

Rytautas replied:
URLs renamed from camelCase to kebab-case

grkarolis commented on src/main/java/com/project/controller/PasswordResetController.java:
```diff
@@ -0,0 +1,65 @@
+        return new GenericResponse("Išsiųstas elektroninis laiškas su nuoroda pasikeisti slaptažodį");
+    }
+
+    @PostMapping("/savePassword")
```
Please rename your URI to something like `save-password` because this does not comply with Open API guidelines.

Rytautas replied:
URLs renamed from camelCase to kebab-case

grkarolis commented on src/main/java/com/project/controller/PasswordResetController.java:
```diff
@@ -0,0 +1,65 @@
+        this.passwordResetService = passwordResetService;
+    }
+
+    @GetMapping("/changePassword")
```
Please rename your URI to something like `change-password` because this does not comply with Open API guidelines.

Rytautas replied:
URLs renamed from camelCase to kebab-case

grkarolis commented on src/main/java/com/project/controller/PasswordResetController.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @GetMapping("/changePassword")
+    public String showChangePasswordPage(Model model, @RequestParam("id") long id, @RequestParam("token") String token) {
```
It looks to me that this service is more for validating the password reset token rather than showing some page. I would suggest returning just an empty response if token was validated successfully and throw an error otherwise.

Rytautas replied:
this service validates the token and, depending on whether its valid or not, returns the path to which the user should be redirected.

grkarolis commented on src/main/java/com/project/controller/PasswordResetController.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @PostMapping("/savePassword")
+    public GenericResponse savePassword(@Valid PasswordResetRequest passwordResetRequest) {
```
Missing request body annotation

Rytautas replied:
added `@RequestBody` annotation

grkarolis commented on src/main/java/com/project/model/Comment.java:
```diff
@@ -0,0 +1,57 @@
+    private boolean isVisibleToApplicant;
+
+    @Column(name = "attachment_name")
+    private String attachmentName;
```
Since the project is about to end at least for future reference I would suggest storing attachment information on a separate table. Your table objects are getting bigger and bigger as you add more data. Try to categorize your data and see what needs a separate table and what data should belong to which data model you have designed.

grkarolis commented on src/main/java/com/project/model/PasswordResetToken.java:
```diff
@@ -0,0 +1,34 @@
+@AllArgsConstructor
+public class PasswordResetToken {
+
+    private static final int EXPIRATION = 60 * 24;
```
This should not be exposed here. Please add to properties file like you did with JWT information.

Rytautas committed:
Feat/password reset (#49)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

grkarolis commented on src/main/java/com/project/service/ApplicationServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+        if (newStatus.equals(ApplicationStatus.PRIIMTA)) {
+            emailService.sendEmail(
+                    app.getApplicant().getEmail(), "Jūsų paraiškos statusas pakeistas",
+                    "Labas!\n" +
```
Store this text somewhere in properties because it does not make any sense to have it here.

Rytautas replied:
stored plain text in `messages.properties`

grkarolis commented on src/main/java/com/project/service/CommentServiceImpl.java:
```diff
@@ -0,0 +1,141 @@
+            app.setLastExternalCommentAuthor(comment.getAuthorEmail());
+            if (author.getRoles().contains(UserRole.ADMIN.name())) {
+                emailService.sendEmail(
+                        app.getApplicant().getEmail(), "Gavote naują komentarą",
```
This can be moved to a properties file since this is just a plain text.

Rytautas replied:
stored plain text in `messages.properties`

grkarolis commented on src/main/java/com/project/service/EmailServiceImpl.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public void constructResetTokenEmail(String contextPath, String token, User user) {
+        String url = contextPath + "/user/changePassword?id=" + user.getId() + "&token=" + token;
+        String message = "Norėdami pasikeisti slaptažodį, paspauskite žemiau esančią nuorodą";
```
Properties file as well.

Rytautas replied:
stored plain text in `messages.properties`

grkarolis commented on src/main/java/com/project/service/EmailServiceImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public void constructResetTokenEmail(String contextPath, String token, User user) {
+        String url = contextPath + "/user/changePassword?id=" + user.getId() + "&token=" + token;
```
String format.

Rytautas replied:
repaced string addition to `String.format`

grkarolis commented on src/main/java/com/project/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public void createPasswordResetTokenForUser(User user, String token) {
+        PasswordResetToken myToken = new PasswordResetToken(token, user);
```
Can it be just something like `resetToken` and not `myToken`?

Rytautas replied:
renamed variable from `myToken` to `resetToken`

grkarolis commented on src/test/java/com/project/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Test
+    public void shouldSucceedInUpdatingUser() throws Exception {
+        //given
```
Why these comments are required here?

Rytautas replied:
removed the comments from test classes

Rytautas committed:
Feat/password reset (#50)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

Rytautas committed:
Feat/password reset (#51)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

Rytautas committed:
Feat/password reset (#52)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

* fixed password not being received from request body

Rytautas committed:
Feat/password reset (#53)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

* fixed password not being received from request body

* fixed password not being received from request body

* changed url sent to the user

Rytautas committed:
Feat/password reset (#54)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

* fixed password not being received from request body

* fixed password not being received from request body

* changed url sent to the user

* changed url sent to the user, hopefully good this time

Rytautas committed:
Feat/password reset (#55)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

* fixed password not being received from request body

* fixed password not being received from request body

* changed url sent to the user

* changed url sent to the user, hopefully good this time

* changed url sent to the user

Rytautas committed:
Feat/password reset (#56)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

* fixed password not being received from request body

* fixed password not being received from request body

* changed url sent to the user

* changed url sent to the user, hopefully good this time

* changed url sent to the user

* some minor changes to password reset
added messages.properties file to store all messages

Rytautas committed:
Feat/password reset (#57)

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* added password reset controller to manage password reset tokens
fixed one test failing due to changes with notifications

* text changes

* replaced generic response to string
created database table to store password reset tokens

* replaced generic response to string
created database table to store password reset tokens

* fixed password not being received from request body

* fixed password not being received from request body

* fixed password not being received from request body

* changed url sent to the user

* changed url sent to the user, hopefully good this time

* changed url sent to the user

* some minor changes to password reset
added messages.properties file to store all messages

* unit test fixes for new changes

mildapetr committed:
changing some dummy data (#58)
